### ISSUE TITLE
Combination of SC Consul, SC Bus and SC Config (1.0.0/1.1.0-RC2) causes double registration on Consul

### ISSUE THREAD

**Issue by donovanmuller** [spring-cloud/spring-cloud-consul#180]
Using the above combination (all 1.0.0/1.1.0-RC2 versions) in an app where config server is looked up via service discovery (`spring.cloud.config.discovery.enabled=true`) causes the app to be registered *twice* in Consul when SC Bus is also enabled (`spring.cloud.bus.enabled=true`):

```
...
2016-05-06 22:23:54.086  INFO 92515 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2016-05-06 22:23:54.087  INFO 92515 --- [           main] o.s.c.consul.discovery.ConsulLifecycle   : Registering service with consul: NewService{id='consul-bus-double-1324109920', name='consul-bus-double', tags=[], address='192.168.1.6', port=8080, check=Check{script='null', interval=10s, ttl=null, http=http://192.168.1.6:8080/health, timeout=null}}
2016-05-06 22:23:54.104  INFO 92515 --- [           main] o.s.c.consul.discovery.ConsulLifecycle   : Registering service with consul: NewService{id='consul-bus-double-912460132', name='consul-bus-double', tags=[], address='192.168.1.6', port=8080, check=Check{script='null', interval=10s, ttl=null, http=http://192.168.1.6:8080/health, timeout=null}}
2016-05-06 22:23:54.110  INFO 92515 --- [           main] io.switchbit.Application                 : Started Application in 18.38 seconds (JVM running for 18.759)
...
```

<img width="1195" alt="screen shot 2016-05-06 at 10 52 12 pm" src="https://cloud.githubusercontent.com/assets/1175891/15085874/3d985330-13dd-11e6-9b4b-de10823192ea.png">

Example app here: https://gist.github.com/donovanmuller/ec3fab8be0cc5f10f06a160abb04e3d9

Basically, if both `spring.cloud.config.discovery.enabled=true` and `spring.cloud.bus.enabled=true` then there is a double registration. My guess is that there are 2 contexts started (based on the different app Id's in the log snippet above (`consul-bus-double-1324109920` and `consul-bus-double-912460132`).
If either one is set to `false`, then there is only a single registration as expected.

Wasn't quite sure where to log this issue as it involves two other SC projects but since the main side effect is double Consul registrations, it seems appropriate here. Feel free to move to other projects if need be...

**Comment by spencergibb** [comment]
My guess is it has to do with `instanceId: ${spring.application.name}-${random.int}`.  Can you not use `${random.int}`. The default now is good enough for running on a single machine since it uses `server.port`. Is this something you're seeing in production (or similar env) or on a laptop?

**Comment by spencergibb** [comment]
And having the bus involved is weird since it has nothing to do with registration. Config discovery I can understand.

**Comment by donovanmuller** [comment]
If I remove `${random.int}` (good to know I don't need it anymore) it still registers twice:

```
spring:
  application:
    name: ${project.artifactId}
  cloud:
    consul:
      enabled: true
      discovery:
        instanceId: ${spring.application.name}
...
```

this happens on my laptop and in a production OpenShift environment

**Comment by spencergibb** [comment]
How can it register twice? I can see it registering with the same id. Also, you can move all of the stream and bus stuff out of bootstrap as it doesn't need to be there. Only information needed for configuration (and discovery if `config.discovery.enabled=true`).

**Comment by donovanmuller** [comment]
I have updated the example app as per your comments here and [here](https://gist.github.com/donovanmuller/ec3fab8be0cc5f10f06a160abb04e3d9#gistcomment-1771124): https://gist.github.com/donovanmuller/ec3fab8be0cc5f10f06a160abb04e3d9/revisions

> How can it register twice?

Even with the updates as per your comments, if `spring.cloud.config.discovery.enabled=true` and `spring.cloud.bus.enabled=true`, the following is logged:

```
2016-05-06 23:20:53.203  INFO 93937 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2016-05-06 23:20:53.206  INFO 93937 --- [           main] o.s.c.consul.discovery.ConsulLifecycle   : Registering service with consul: NewService{id='consul-bus-double', name='consul-bus-double', tags=[], address='192.168.1.6', port=8080, check=Check{script='null', interval=10s, ttl=null, http=http://192.168.1.6:8080/health, timeout=null}}
2016-05-06 23:20:53.222  INFO 93937 --- [           main] o.s.c.consul.discovery.ConsulLifecycle   : Registering service with consul: NewService{id='consul-bus-double', name='consul-bus-double', tags=[], address='192.168.1.6', port=8080, check=Check{script='null', interval=10s, ttl=null, http=http://192.168.1.6:8080/health, timeout=null}}
2016-05-06 23:20:53.228  INFO 93937 --- [           main] io.switchbit.Application                 : Started Application in 7.717 seconds (JVM running for 8.257)
```

I.e. 2 `Registering service with consul`s
This is what I mean when I say two registrations. 
As per the app logs above and Consul node logs below (running Docker container):

```
$ docker run --rm -it -p 8500:8500 -p 8400:8400 -p 8600/udp -h node1 gliderlabs/consul-server -bootstrap -advertise $(docker-machine ip dev) -client 0.0.0.0
==> WARNING: Bootstrap mode enabled! Do not enable unless necessary
==> WARNING: It is highly recommended to set GOMAXPROCS higher than 1
==> Starting Consul agent...
==> Starting Consul agent RPC...
==> Consul agent running!
         Node name: 'node1'
        Datacenter: 'dc1'
            Server: true (bootstrap: true)
       Client Addr: 0.0.0.0 (HTTP: 8500, HTTPS: -1, DNS: 8600, RPC: 8400)
      Cluster Addr: 192.168.99.100 (LAN: 8301, WAN: 8302)
    Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false
             Atlas: <disabled>

==> Log data will now stream in as it occurs:

    2016/05/06 21:28:16 [INFO] serf: EventMemberJoin: node1 192.168.99.100
    2016/05/06 21:28:16 [INFO] serf: EventMemberJoin: node1.dc1 192.168.99.100
    2016/05/06 21:28:16 [INFO] raft: Node at 192.168.99.100:8300 [Follower] entering Follower state
    2016/05/06 21:28:16 [INFO] consul: adding server node1 (Addr: 192.168.99.100:8300) (DC: dc1)
    2016/05/06 21:28:16 [INFO] consul: adding server node1.dc1 (Addr: 192.168.99.100:8300) (DC: dc1)
    2016/05/06 21:28:16 [ERR] agent: failed to sync remote state: No cluster leader
    2016/05/06 21:28:18 [WARN] raft: Heartbeat timeout reached, starting election
    2016/05/06 21:28:18 [INFO] raft: Node at 192.168.99.100:8300 [Candidate] entering Candidate state
    2016/05/06 21:28:18 [INFO] raft: Election won. Tally: 1
    2016/05/06 21:28:18 [INFO] raft: Node at 192.168.99.100:8300 [Leader] entering Leader state
    2016/05/06 21:28:18 [INFO] consul: cluster leadership acquired
    2016/05/06 21:28:18 [INFO] consul: New leader elected: node1
    2016/05/06 21:28:18 [INFO] raft: Disabling EnableSingleNode (bootstrap)
    2016/05/06 21:28:18 [INFO] consul: member 'node1' joined, marking health alive
    2016/05/06 21:28:18 [INFO] agent: Synced service 'consul'

    2016/05/06 21:28:28 [INFO] agent: Synced service 'config-server-886420342'
    2016/05/06 21:28:36 [INFO] agent: Synced check 'service:config-server-886420342'

    2016/05/06 21:28:44 [INFO] agent: Synced service 'consul-bus-double' # Registration 1
    2016/05/06 21:28:44 [INFO] agent: Synced service 'consul-bus-double' # Registration 2

    2016/05/06 21:28:50 [INFO] agent: Synced check 'service:consul-bus-double'
```

there are 2 distinct calls to register the app, instead of the expected one.
It might not seem that way if you look at the UI now:

<img width="1198" alt="screen shot 2016-05-06 at 11 31 41 pm" src="https://cloud.githubusercontent.com/assets/1175891/15086684/b94d51ce-13e2-11e6-834e-a23a77d9d9fc.png">

because since removing `-${random.int}` from `instanceId `, the app Id is now the same *and* the port is also the same (because it's the same app) so it seems like only one registration but in reality there were two registrations, you just can't see that.

**Comment by spencergibb** [comment]
Yes, that is the expected behavior. Until we separate out registration from discovery, to lookup config via discovery we have to register in the bootstrap context and then it happens again in the normal context.

**Comment by donovanmuller** [comment]
Ok, that makes sense. 
I guess the issue for me then was really the use of `-${random.int}` in `instanceId`.
This caused Consul to have two nodes registered for the same app instance because of this re-registering (getting a different `random.int` at each context).

So the only question I still have is why the re-registering *doesn't happen* when I set `spring.cloud.bus.enabled=false`?
There is only a single log for `Registering service with consul` and call to Consul:

```
2016-05-06 23:52:53.844  INFO 94656 --- [           main] s.b.c.e.t.TomcatEmbeddedServletContainer : Tomcat started on port(s): 8080 (http)
2016-05-06 23:52:53.845  INFO 94656 --- [           main] o.s.c.consul.discovery.ConsulLifecycle   : Registering service with consul: NewService{id='consul-bus-double', name='consul-bus-double', tags=[], address='192.168.1.6', port=8080, check=Check{script='null', interval=10s, ttl=null, http=http://192.168.1.6:8080/health, timeout=null}}
2016-05-06 23:52:53.862  INFO 94656 --- [           main] io.switchbit.Application                 : Started Application in 5.207 seconds (JVM running for 5.615)
```

```
    ...
    2016/05/06 21:52:53 [INFO] agent: Synced service 'consul-bus-double'
```

as opposed to the double log/call shown above.
Is it not re-registering in this case? Discovering the config server still works...

**Comment by spencergibb** [comment]
That, I haven't the slightest idea.

**Comment by spencergibb** [comment]
My guess is that it causes some kind of eager bean creation.

**Comment by donovanmuller** [comment]
👻

**Comment by donovanmuller** [comment]
@spencergibb Could be related to https://github.com/spring-cloud/spring-cloud-consul/issues/180#issuecomment-217570740: https://github.com/spring-projects/spring-integration-java-dsl/issues/80#issuecomment-218245617?

**Comment by spencergibb** [comment]
Could be
